### PR TITLE
[FLINK-22096][tests] Fix port conflict for ServerTransportErrorHandlingTest#testRemoteClose

### PR DESCRIPTION
**Pull request by TanYuxin-tyx**
## What is the purpose of the change

Fix port conflict in `ServerTransportErrorHandlingTest#testRemoteClose`. In the test case, a `BindException`  may be thrown when init Netty server. When initializing Netty server, `NetUtils.getAvailablePort()` is called in `createConfig()`. After obtaining the available port, the port may be used by other processes, which may lead to `BindException`. To resolve the issue, a retry is added when calling `initServerAndClient` method.

## Brief change log
  - *Add the number of retries when initializing the Netty server in the test case*

## Verifying this change
  - *Initializing  the Netty server successfully in the test case*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/ServerTransportErrorHandlingTest.java [hunk 1]
```diff
@@ -101,7 +104,22 @@ public void channelRead(ChannelHandlerContext ctx, Object msg)
...
 
         try {
-            serverAndClient = initServerAndClient(protocol, createConfig());
+            for (int retry = 0; retry < NETTY_INIT_MAX_RETRY_TIMES; retry++) {
+                try {
+                    serverAndClient = initServerAndClient(protocol, createConfig());
+                    break;
+                } catch (Exception e) {
+                    if (retry >= NETTY_INIT_MAX_RETRY_TIMES - 1) {
+                        throw e;
```
Comment (guoweiM) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/ServerTransportErrorHandlingTest.java:113]: Maybe we should give more details in the `Exception`. Such as the retry failed. WDYT?

Comment (TanYuxin-tyx) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/ServerTransportErrorHandlingTest.java:113]: Thanks a lot for reviewing. @guoweiM   ok，I will fix it recently.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/ServerTransportErrorHandlingTest.java [hunk 2]
```diff
@@ -101,7 +104,22 @@ public void channelRead(ChannelHandlerContext ctx, Object msg)
...
+                    }
+                    if (e instanceof BindException
```
Comment (guoweiM) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/ServerTransportErrorHandlingTest.java:115]: I am not sure but from the test url I find the `Exception` is following
image

So I would suggest to see what exactly exception is when the port is used.
WDYT?

Comment (TanYuxin-tyx) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/ServerTransportErrorHandlingTest.java]: Thanks. I will make the exception inevitable by init the server using an occupied port. And I will print the detailed stack of the exception by triggering it in the CI test.

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

**Comment by wsry** [comment]
@flinkbot run azure

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

**Comment by TanYuxin-tyx** [comment]
As shown above, I have reproduced the above exceptions. I added a new piece of code to print this exception.
```java
                System.out.println("e instanceof BindException? " + (e instanceof BindException));
                System.out.println("e.toString(): " + e.toString());
                System.out.println("e.getLocalizedMessage(): " + e.getLocalizedMessage());
                System.out.println(
                        "e contains Address already in use? "
                                + (ExceptionUtils.findThrowableWithMessage(
                                                e, "Address already in use"))
                                        .isPresent());
```
The test results are here. 
https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=25907&view=logs&j=a57e0635-3fad-5b08-57c7-a4142d7d6fa9&t=2ef0effc-1da1-50e5-c2bd-aab434b1c5b7&l=6530

As the test results above, I found the `BindException` will be wrapped into `NativeIoException`. The content of `NativeIoException` is 'Address already in use'. 
image

However, when running on the local Mac, the exception thrown by the same code is `BindException`. This is a screenshot from my local IDEA test on Mac. 
image

In my opinion, the exception thrown initially is `BindException`, but when running the test on the test platform, it is wrapped as `NativeIoException`.
So I think catch both `BindException` and exceptions with 'Address already in use' are necessary. 
WDYT?, @guoweiM, @wsry. Thanks again.

**Comment by guoweiM** [comment]
Thanks @TanYuxin-tyx for invest this problem. I looks good to me. Please go ahead

**Comment by wsry** [comment]
@flinkbot run azure

**Comment by wsry** [comment]
I will help to merge this PR after all tests pass.

**Review by wsry — APPROVED** [review]

**Comment by wsry** [comment]
Merged.

**Pull request closed without merging**